### PR TITLE
feat(#188): CLI integration — query execution, auto-index, staleness detection, git hooks

### PR DESCRIPTION
**Pull request by dean0x**
## Summary

- `skim search "auth"` previously printed "not yet implemented" — this PR wires all search components together so the command executes BM25F search, auto-builds the index on first use, and refreshes when the git HEAD changes
- Adds 5 new modules (staleness, snippet, query, hooks, query/hooks tests) and extends manifest, types, index, and CLI dispatch
- 56 new tests across all new modules; all 2,429+ existing tests continue to pass

### New modules

| Module | What it does |
|--------|-------------|
| `staleness.rs` | Reads git HEAD (loose refs, packed-refs, worktree `.git` file) and classifies index as `Current \| HeadChanged \| NoStoredHead \| NoIndex` |
| `snippet.rs` | Maps byte offsets to line numbers, extracts N-line context windows, validates against manifest mtime |
| `query.rs` | `execute_query` → auto-build → BM25F search → `FileId→path` resolution → snippet extraction → `QueryOutput` |
| `hooks.rs` | Installs/removes marker-delimited `skim search --update` block in `post-commit`, `post-merge`, `post-checkout` (idempotent) |

### Extended files

- **`manifest.rs`** — `git_head: Option<String>` on `ManifestHeader` with `#[serde(default)]` for backward compat; `sorted_paths()`, `entry_count()`, `stored_git_head()`, `set_git_head()` methods
- **`index.rs`** — stores current git HEAD in manifest on every build; `resolve_search_cache_dir` made `pub(super)`
- **`types.rs`** — `QueryConfig`, `ResolvedResult`, `QueryOutput`, `SnippetContext`, `SnippetLine` types
- **`mod.rs`** — full CLI dispatch: `--build` / `--rebuild` / `--update` / `--stats` / `--install-hooks` / `--remove-hooks` / `--json` / `--limit N` / bare query text
- **`init/install.rs`** — installs search hooks + spawns background index build on `skim init`

## Breaking Changes

`ManifestHeader` gains an optional `git_head` field. Old manifests (without this field) deserialize cleanly with `git_head: None` via `#[serde(default)]` — they trigger a one-time refresh on the next query, then record the HEAD.

## Reviewer Focus Areas

- **`FileId → path` invariant** (`query.rs:resolve_paths_and_snippets`): relies on `sorted_paths()[n]` = path for `FileId(n)`. The walk pipeline sorts entries and the consumer assigns sequential IDs — this invariant is documented in both `manifest.rs` and `walk.rs`
- **Snippet mtime guard** (`snippet.rs:extract_snippet`): when manifest entry has a stored mtime and it differs from the file's current mtime, `None` is returned instead of serving stale byte offsets
- **Staleness detection** (`staleness.rs`): pure file I/O, no subprocess — handles detached HEAD (40-hex SHA), symbolic refs (`ref: refs/heads/main`), loose ref files, `packed-refs` fallback, and worktree `.git` pointer files
- **Git hook idempotency** (`hooks.rs:install_one_hook`): checks for `# skim-search-start` marker before writing; running install twice produces exactly one block

### CODE REVIEW

**Comment by dean0x** [comment]
## Code Review Summary

This PR adds comprehensive CLI search integration to skim with 2,341 new lines across 14 files. The core design is sound (streaming pipeline, bounded channels, atomic writes), and the feature is well-tested with 56 new tests covering 5 test modules.

**Critical Issues Found (≥80% confidence):** 7 blocking issues require fixes before merge.

### Blocking Issues

#### 1. Infinite Rebuild Loop for Non-Git Projects
**Files:** `staleness.rs` (lines 153-186, 199-252)
**Severity:** HIGH (Performance Regression)
**Confidence:** 95% (Architecture) + 82% (Regression)

When `read_git_head()` returns `None` (non-git project), `build_index` stores `git_head: None`. On the next query, `check_staleness` returns `NoStoredHead`, triggering a full rebuild. Every query triggers another rebuild—O(project_size) per search. This breaks non-git projects.

**Fix:** If both stored and current HEAD are `None`, return `StalenessCheck::Current` instead of `NoStoredHead`.

#### 2. Predictable Temp File Path in `write_hook_atomic`
**Files:** `hooks.rs` (line 176)
**Severity:** HIGH (Security)
**Confidence:** 80% (Security) + 82% (Reliability)

Uses deterministic path `hook_path.with_extension("tmp")`. A local attacker could place a symlink at `.git/hooks/post-commit.tmp` before the write, redirecting content to a target of their choice. The manifest module correctly uses `tempfile::NamedTempFile`—hooks should follow the same pattern.

#### 3. Duplicate Manifest Load on Query Hot Path
**Files:** `query.rs` (lines 55, 70)
**Severity:** HIGH (Performance)
**Confidence:** 92%

`execute_query` calls `auto_refresh_if_stale()` (loads manifest), then loads it again to build sorted paths. Parses the entire manifest twice per query. On 10K files, this doubles I/O and JSON parse cost.

**Fix:** Return the loaded manifest from `auto_refresh_if_stale` so the caller can reuse it.

#### 4. No File-Size Guard on Snippet Extraction
**Files:** `snippet.rs` (line 127)
**Severity:** HIGH (Reliability)
**Confidence:** 90%

`extract_snippet` reads entire files into memory without size check. While indexing caps at 5 MB, extraction bypasses this. A file grown since indexing could exhaust memory across 20 result reads.

#### 5. No Concurrent Index Build Protection
**Files:** `install.rs` (lines 316-324)
**Severity:** HIGH (Reliability)
**Confidence:** 85%

`execute_install` spawns background build, but `auto_refresh_if_stale` and git hooks can trigger concurrent builds. Multiple builders write to the same index files without coordination, risking corruption.

**Fix:** Add file lock (e.g., `.skim-build.lock`) before building.

#### 6. Unrecognized Flags Absorbed Into Query Text
**Files:** `mod.rs` (lines 158-159)
**Severity:** HIGH (User Experience)
**Confidence:** 85% (Rust) + 85% (Architecture)

`parse_flags` treats flag typos like `--buld` or `--jsno` as query text. No error message—searches for the misspelled flag instead.

#### 7. parse_flags Returns Bare Flags Instead of Result
**Files:** `mod.rs` (line 116)
**Severity:** HIGH (Consistency)
**Confidence:** 85%

Deviates from sibling subcommands which return `anyhow::Result<T>`. Silently accepts invalid `--limit abc` (keeps default) instead of reporting an error. Pattern mismatch with heatmap and init modules.

**Fix:** Return `anyhow::Result<Flags>` with proper error messages.

---

## Lower-Confidence Suggestions (60-79%)

**Flagged by multiple reviewers:**

1. **`is_hex_sha` rejects SHA-256 (70-85%):** Git repos using SHA-256 would silently degrade to rebuild-on-every-query. Accept both 40-char and 64-char hex strings.

2. **FileId → path invariant needs enforcement (82%):** The mapping `sorted_paths()[n] == path for FileId(n)` is implicit. Any future change disrupting sort order silently corrupts results. Add debug assertion or store FileId explicitly in manifest entries.

3. **Hand-rolled flag parsing vs clap (65-85%):** The `index` subcommand uses clap, while `search` uses manual parsing. Inconsistent error handling and help text generation.

4. **`extract_context_window` collects all lines (65-82%):** Allocates Vec for entire file, but only uses 7 lines (~3 context + 1 match + 3 context). Use iterator-based approach for large files.

5. **Misleading comment in staleness.rs (85%):** Line 174 comment says "treat as NoStoredHead if manifest also has no HEAD" but the manifest DOES have stored HEAD at that point. Comment is contradictory.

6. **`run_stats` mixed concerns (85% Complexity):** Function mixes index existence check, data gathering, and formatting (JSON vs text). Extract formatting into helpers.

7. **Orphaned background process (82% Reliability):** The spawned `skim search --build` Child is dropped without wait(), leaving zombie on Unix. Add comment explaining intentional fire-and-forget, or use explicit `forget()`.

---

## Overall Assessment

**Architecture:** Clean module decomposition with clear responsibilities. Streaming pipeline with bounded channels is well-designed.

**Test Coverage:** 56 new tests across 5 modules. Good coverage of happy paths, edge cases tested at function level.

**Reliability:** Bounded iteration everywhere, atomic writes with temp-file-rename, graceful degradation via outcome enums. Strong fundamentals, but concurrent build race and unbounded snippet extraction are concerning.

**Security:** Proper SQL parameterization (N/A), validated SHA outputs, hardcoded hook content. Symlink race in write_hook_atomic is the actionable item.

**Performance:** Well-optimized streaming design. Duplicate manifest load is a clear win if refactored.

---

## Merge Recommendation: **CHANGES_REQUESTED**

The 7 blocking issues must be resolved before merge:
- **2 HIGH severity** (non-git rebuild loop, concurrent build race) are functional regressions
- **3 HIGH severity** (symlink race, double manifest load, unbounded snippet) affect security/reliability
- **2 HIGH severity** (flag handling) affect user experience and consistency

After fixes, this is a solid 2.3K-line addition with strong fundamentals. The lower-confidence suggestions (60-79%) should be addressed in follow-up refinement.

File: crates/rskim/src/cmd/search/staleness.rs [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+    // Read current HEAD.
+    let current = match read_git_head(project_root) {
+        Some(h) => h,
+        None => {
+            // Non-git project or can't read HEAD — treat as NoStoredHead if
+            // the manifest also has no HEAD stored.
+            return StalenessCheck::NoStoredHead;
+        }
+    };
+
```
Comment (dean0x) [inline crates/rskim/src/cmd/search/staleness.rs:231]: **Infinite rebuild loop for non-git projects** (95% confidence)

When `read_git_head()` returns `None` (non-git), `build_index` stores `git_head: None`. On next query, `check_staleness` returns `NoStoredHead` → unconditional rebuild. Every query triggers rebuild. O(project_size) per search.

**Fix:** If both `stored_git_head()` and `read_git_head()` are `None`, return `StalenessCheck::Current` instead of `NoStoredHead`.

**Review by dean0x — COMMENTED** [review]
See summary comment for detailed findings.

**Comment by dean0x** [comment]
## PR Comments Created

### Summary Comment
A comprehensive review comment has been posted with detailed findings, scores, and recommendations.

### Inline Comments Posted
1. **crates/rskim/src/cmd/search/staleness.rs (line 180)** - Infinite rebuild loop for non-git projects (95% confidence)

### Additional Blocking Issues (Comprehensive Details in Summary)

The full review identified 7 blocking issues requiring fixes:

1. **Infinite rebuild loop for non-git projects** (staleness.rs:153-186, 199-252) — Architecture 95%, Regression 82%
   - When `read_git_head()` returns `None`, stored HEAD is also `None`, causing `NoStoredHead` return → rebuild every query
   - Fix: Return `StalenessCheck::Current` when both stored and current HEAD are `None`

2. **Predictable temp file path in hooks.rs** (line 176) — Security 80%, Reliability 82%
   - Symlink race: deterministic `.tmp` extension allows local attacker to hijack hook
   - Fix: Use `tempfile::NamedTempFile` like the manifest module does

3. **Duplicate manifest load on query hot path** (query.rs:55,70) — Performance 92%
   - Manifest loaded twice per query: once in `auto_refresh_if_stale()`, again for sorted paths
   - Fix: Return manifest from `auto_refresh_if_stale` for reuse

4. **No file-size guard on snippet extraction** (snippet.rs:127) — Reliability 90%
   - `extract_snippet` reads entire files without size cap; could exhaust memory on large/grown files
   - Fix: Add metadata size check matching the 5 MB indexing cap

5. **No concurrent index build protection** (install.rs:316) — Reliability 85%
   - Background build from `execute_install` can race with `auto_refresh_if_stale` and git hooks
   - Fix: Add advisory file lock (`.skim-build.lock`) before building

6. **Unrecognized flags absorbed as query text** (mod.rs:158) — Rust 85%, Architecture 85%
   - Typos like `--buld` become search queries with no error
   - Fix: Validate `--` flags and error on unrecognized ones

7. **parse_flags return type inconsistent** (mod.rs:116) — Consistency 85%
   - Returns bare `Flags` while siblings return `anyhow::Result<T>`
   - Silently accepts `--limit abc` instead of reporting error
   - Fix: Return `anyhow::Result<Flags>` with validation

---

## Review Statistics

- **Files reviewed:** 14 files (code + tests)
- **Lines added:** 2,341
- **Test coverage:** 56 new tests across 5 test modules
- **Blockers identified:** 7 (all ≥80% confidence)
- **Lower-confidence suggestions:** 7 (60-79% confidence)

### Reviewer Coverage
- Security Review: 1 MEDIUM blocking (predictable temp), 2 MEDIUM should-fix (ref validation, --limit cap)
- Architecture Review: 2 HIGH blocking, 2 MEDIUM should-fix
- Performance Review: 2 HIGH blocking, 2 MEDIUM should-fix
- Complexity Review: 2 HIGH, 2 MEDIUM
- Rust Review: 3 HIGH, 2 MEDIUM
- Reliability Review: 3 HIGH, 2 MEDIUM
- Consistency Review: 2 HIGH, 2 MEDIUM
- Regression Review: 1 HIGH, 1 MEDIUM
- Testing Review: 2 HIGH, 2 MEDIUM

---

## Merge Recommendation

**CHANGES_REQUESTED**

All 7 blocking issues must be resolved before merge. The two functional regressions (infinite rebuild loop for non-git, concurrent build race) are the most critical. After fixes, this is a solid 2.3K-line addition with strong fundamentals.

---

*Code Review completed by automated multi-reviewer analysis with Claude Code attribution.*

File: crates/rskim/src/cmd/search/mod.rs [hunk 2]
```diff
@@ -43,9 +58,305 @@ pub(crate) fn run(
...
+    let mut i = 0;
+    while i < args.len() {
+        match args[i].as_str() {
+            "--build" => action_flag = Some(SearchAction::Build),
+            "--rebuild" => action_flag = Some(SearchAction::Rebuild),
+            "--update" => action_flag = Some(SearchAction::Update),
+            "--stats" => action_flag = Some(SearchAction::Stats),
+            "--install-hooks" => action_flag = Some(SearchAction::InstallHooks),
+            "--remove-hooks" => action_flag = Some(SearchAction::RemoveHooks),
+            "--json" => json = true,
```
Comment (dean0x) [inline crates/rskim/src/cmd/search/mod.rs]: **BREAKING CHANGE**: The `-j` short alias for `--json` has been removed.

Previously: `"--json" | "-j" => json = true`
Currently: `"--json" => json = true`

This breaks backward compatibility for scripts using `skim search -j`. The new error handling will reject `-j` as unrecognized.

**Fix**: Restore the short alias:
```rust
"--json" | "-j" => json = true,
```

Also add `-j` to the valid flags list in the error message at line 171.

**Confidence**: 95% (Consistency + Regression reviewers)

File: crates/rskim/src/cmd/search/snippet.rs [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+            .map(|d| d.as_secs());
+        if current_mtime != Some(stored_mtime) {
+            return SnippetOutcome::Stale;
+        }
+    }
+
+    // Size guard: reject files larger than 5 MB to match the index-build cap and
+    // bound peak memory when 20 results are resolved simultaneously.
+    const MAX_SNIPPET_FILE_BYTES: u64 = 5 * 1024 * 1024;
+    let file_size = std::fs::metadata(&abs_path)
```
Comment (dean0x) [inline crates/rskim/src/cmd/search/snippet.rs]: **PERFORMANCE**: Duplicate `std::fs::metadata()` syscalls in `extract_snippet`.

Lines 124 and 137 both call `std::fs::metadata(&abs_path)` on the same file:
- Line 124: mtime guard calls `metadata()`
- Line 137: size guard calls `metadata()` again

With `--limit 20` (default), this adds 20 extra `stat(2)` syscalls per query. On NFS or high-latency filesystems, this measurably slows down snippet extraction.

**Fix**: Hoist a single metadata call and share between both guards:
```rust
let meta = std::fs::metadata(&abs_path).ok();

// Mtime guard
if let Some(stored_mtime) = manifest_entry.and_then(|e| e.mtime) {
    let current_mtime = meta.as_ref()
        .and_then(|m| m.modified().ok())
        .and_then(|t| t.duration_since(std::time::UNIX_EPOCH).ok())
        .map(|d| d.as_secs());
    if current_mtime != Some(stored_mtime) {
        return SnippetOutcome::Stale;
    }
}

// Size guard
const MAX_SNIPPET_FILE_BYTES: u64 = 5 * 1024 * 1024;
let file_size = meta.as_ref().map(|m| m.len()).unwrap_or(0);
if file_size > MAX_SNIPPET_FILE_BYTES {
    return SnippetOutcome::Unavailable;
}
```

**Confidence**: 90% (Performance), 88% (Rust), 80% (Consistency)

File: crates/rskim/src/cmd/search/staleness.rs [hunk 4]
```diff
@@ -0,0 +1,318 @@
...
+}
+
+impl std::fmt::Display for StalenessCheck {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            StalenessCheck::Current => write!(f, "current"),
+            StalenessCheck::HeadChanged { stored, current } => write!(
+                f,
+                "stale (HEAD changed: {}…→{}…)",
+                &stored[..8.min(stored.len())],
```
Comment (dean0x) [inline crates/rskim/src/cmd/search/staleness.rs]: **RELIABILITY**: `StalenessCheck::Display` string slicing could panic on non-ASCII input.

Lines 43-44: `&stored[..8.min(stored.len())]`

While SHAs are normally ASCII, the `stored` and `current` strings come from `read_git_head`, which reads arbitrary file content. A corrupted `.git/HEAD` could produce non-ASCII that passes validation. Slicing at byte position 8 could split a multi-byte UTF-8 codepoint and panic at runtime.

**Impact**: CLI crashes when displaying staleness for corrupted repos.

**Fix**: Use safe char-aware slicing instead of byte-index slicing:
```rust
let stored_short = stored.chars().take(8).collect::<String>();
let current_short = current.chars().take(8).collect::<String>();
write!(f, "stale (HEAD changed: {stored_short}...→{current_short}...)")?;
```

Or use `.get()` which returns `None` instead of panicking:
```rust
let stored_short = stored.get(..8).unwrap_or(stored);
```

Same issue appears at line 290-291 in `auto_refresh_if_stale`.

**Confidence**: 82% (Reliability)

File: crates/rskim/src/cmd/search/mod.rs [hunk 5]
```diff
@@ -43,9 +58,305 @@ pub(crate) fn run(
...
+
+// ============================================================================
+// Parsed flags
+// ============================================================================
+
+/// The action the user wants to perform, derived from CLI flags.
+///
+/// Encodes the mutually-exclusive mode flags as a single enum variant so that
+/// dispatch is a `match` rather than a cascade of `if flags.X` checks.
+#[derive(Debug, PartialEq)]
```
Comment (dean0x) [inline crates/rskim/src/cmd/search/mod.rs]: **CONSISTENCY**: `SearchAction` enum should derive `Eq` alongside `PartialEq`.

Line 90: `#[derive(Debug, PartialEq)]`

The codebase convention (see `IndexResult` in `types.rs:126`) is to derive both `PartialEq` and `Eq` for types where all variants are `Eq`. `SearchAction` contains only unit variants and `String` (which is `Eq`), so the `Eq` bound is satisfied.

**Fix**:
```rust
#[derive(Debug, PartialEq, Eq)]
enum SearchAction {
```

**Confidence**: 82% (Consistency)

**Comment by dean0x** [comment]
# Code Review Summary — PR #245

## Overall Status: CHANGES REQUESTED

The refactoring from boolean flags to `SearchAction` enum is a strong improvement — it eliminates error-prone cascade patterns and makes invalid states unrepresentable. However, there are **4 blocking issues** that need to be addressed before merge.

---

## Blocking Issues (≥80% confidence)

### HIGH Priority
1. **Removed `-j` short alias for `--json`** (mod.rs:136)
   - Breaking change: scripts using `skim search -j` will now fail
   - Confidence: 95%

2. **String slicing panic risk in `StalenessCheck::Display`** (staleness.rs:43-44)
   - Could crash on corrupted repos with non-ASCII in `.git/HEAD`
   - Confidence: 82%

### MEDIUM Priority
3. **Duplicate `std::fs::metadata()` syscalls** (snippet.rs:124,137)
   - Adds 20 extra stat syscalls per query (with default `--limit 20`)
   - Performance impact on NFS/high-latency filesystems
   - Confidence: 90% (Performance), 88% (Rust), 80% (Consistency)

4. **Missing `Eq` derive on `SearchAction`** (mod.rs:90)
   - Breaks codebase consistency pattern
   - Confidence: 82%

---

## Should-Fix Issues (≥80% confidence)

### MEDIUM Priority
- **`check_staleness` returns bare tuple** (staleness.rs:189)
  - Prone to positional-swap bugs as API evolves
  - Recommend: Introduce `StalenessResult { status, manifest }` struct
  - Confidence: 80% (Architecture)

- **Missing `Display` trait tests** (staleness.rs:36-49)
  - New Display impl is user-facing but untested
  - Recommend: Add tests for each StalenessCheck variant
  - Confidence: 85% (Testing)

- **`parse_flags` lacks conflicting-action test** (mod.rs:120)
  - Last action flag silently wins when multiple specified
  - Behavior should be documented with a test
  - Confidence: 80% (Testing)

- **`--limit 0` accepted silently** (mod.rs:142)
  - Error message says "must be positive" but `0` is accepted
  - Could cause unbounded results if treated as "no limit" downstream
  - Confidence: 85% (Reliability)

- **Conflicting action flags silently lost** (mod.rs:130-135)
  - `skim search --build --stats` silently runs stats (ignores build)
  - Violates "fail loud" principle
  - Confidence: 80% (Reliability)

### LOW Priority
- **Double iteration over lines** (snippet.rs:66,83)
  - `lines().count()` then `lines().skip().take()`
  - Minor inefficiency bounded by 5 MB size guard
  - Confidence: 80% (Performance)

---

## Lower-Confidence Suggestions (60-79%)

1. **Auto-refresh re-loads manifest** (staleness.rs:308) — Could avoid disk round-trip if `build_index` returned manifest (Confidence: 70%)

2. **Hand-rolled parse_flags vs clap derive** (mod.rs:86-189) — Parser duplicates clap but is correct and well-tested (Confidence: 65%)

3. **BTreeMap vs HashMap tradeoff** (manifest.rs:112) — Net win for typical sizes, eliminates two `sort_unstable()` calls (Confidence: 65%)

4. **Advisory lock blocks indefinitely** (index.rs:176) — No timeout on `lock_file.lock()`, could hang on zombie process (Confidence: 60-70%)

---

## Positive Observations

✅ **Excellent enum-driven dispatch** — SearchAction replaces boolean cascade

✅ **Result-returning parse_flags** — Invalid flags now error vs silent fallback

✅ **Sound staleness state machine** — Four-case match eliminates infinite rebuild loop

✅ **Security improvements** — Path traversal guard, TOCTOU fix, advisory locking, input validation

✅ **Strong test coverage** — 56 new tests covering error cases and edge cases

✅ **Clear complexity reduction** — BTreeMap optimization, NamedTempFile RAII, Display impl

---

## Next Steps

1. Address all 4 blocking issues before merge
2. Consider should-fix items as follow-up improvements
3. Lower-confidence suggestions are notes for future refactoring

**Generated by Code Review Agent** — 9 comprehensive reviews: Security, Architecture, Performance, Complexity, Consistency, Regression, Testing, Reliability, Rust
